plugins/catalog-backend-module-msgraph/src/microsoftGraph/read.ts

```diff
@@ -244,6 +245,31 @@ export async function readMicrosoftGraphGroups(
 
           if (member['@odata.type'] === '#microsoft.graph.group') {
             ensureItem(groupMember, group.id!, member.id);
+
+            if (options?.groupIngestMemberGroups) {
+              const groupMemberEntity = await transformer(member);
+
+              if (groupMemberEntity) {
+                groups.push(groupMemberEntity);
+
+                for await (const subMember of client.getGroupMembers(
+                  member.id!,
+                  { top: PAGE_SIZE },
+                )) {
+                  if (!subMember.id) {
+                    continue;
+                  }
+
+                  if (subMember['@odata.type'] === '#microsoft.graph.user') {
+                    ensureItem(groupMemberOf, subMember.id, member.id!);
+                  }
+
+                  if (subMember['@odata.type'] === '#microsoft.graph.group') {
+                    ensureItem(groupMember, member.id!, subMember.id);
+                  }
+                }
+              }
+            }
```

Comment:
`group` is the group we found using the normal filter/search/... criteria.

We go through all its members (users and groups) and establish their relationships.

With the added option, the intent is to add "its sub-groups".

Here, an open question could be, how many levels you want to support. Should this be just one level. Should this be recursive/down to the leaves?

Instead of a boolean, you could use the number of levels as value:
```yaml
group:
  includeSubGroups: 2
  # or
  includeSubGroupLevels: 2
```

Assuming, `group` has another group as a member: why not using the same logic as for the original group (with or without the optional section re sub-groups depending on whether we want to traverse further or keep it at one only.

```ts
function async loadGroup(group: any, includeSubGroups: boolean): Promise<GroupEntity[]> {
  const groups: GroupEntity[] = [];
  const entity = await transform(group);
  if (!entity) {
    return groups;
  }
  
  for await (const member of client.getGroupMembers(group.id!, {
    top: PAGE_SIZE,
  })) {
    if (!member.id) {
      continue;
    }

    if (member['@odata.type'] === '#microsoft.graph.user') {
      ensureItem(groupMemberOf, member.id, group.id!);
      continue;
    }

    if (member['@odata.type'] === '#microsoft.graph.group') {
      ensureItem(groupMember, group.id!, member.id);
      
      if (includeSubGroups) {
        // here, we could decide about if and how often we want to go deeper
        loadGroups(member, false);
      }
    }
}
```

Reply:
Good question.

Thinking about it, I see three alternatives how this feature could work:

** (1) Keep the current semantics, and only include sub groups of level 1.**
- Probably our favourite, since it achieves to implement our use case with the least amount of changes.
- Not necessarily generic enough to apply to different organisational setups and group structures in Azure Entra ID.

** (2) Take the opposite extreme, and always traverse down to the leaves.**
- Not sure this even makes sense.
- Could be quite performance heavy for large organisations with a high degree of nesting (would probably require a hard recursion limit).
- Would probably some research into typical Entra ID group structures of enterprise organisations if valuable at all.

** (3) Make the traversal depth configurable (as per your suggestion).**
- The most generic variant, but probably also bit more complex to implement and test.
- Not sure if worth the effort, I did not come across any related feature/bug requests... do you know of similar requirements?


Our suggestion would be to go with (1) for now, and extend to (3) in the future if the need arises.
What do you think?